redwood-treehouse-composeui/src/commonMain/kotlin/app/cash/redwood/treehouse/composeui/TreehouseContent.kt

```diff
@@ -43,26 +46,25 @@ public fun <A : AppService> TreehouseContent(
   )
 
   val rememberedCodeListener = rememberUpdatedState(codeListener)
-  val rememberedContent = rememberUpdatedState(contentSource)
   val treehouseView = remember(widgetSystem) {
     object : TreehouseView<A> {
       override val codeListener get() = rememberedCodeListener.value
-      override var stateChangeListener: OnStateChangeListener<A>? = null
-      override val boundContentSource: TreehouseContentSource<A> get() = rememberedContent.value
       override val children = ComposeWidgetChildren()
       override val hostConfiguration = MutableStateFlow(hostConfiguration)
       override val widgetSystem = widgetSystem
+      override val readyForContent = true
+      override var readyForContentChangeListener: ReadyForContentChangeListener<A>? = null
       override fun reset() = children.remove(0, children.widgets.size)
     }
   }
   LaunchedEffect(treehouseView, hostConfiguration) {
     treehouseView.hostConfiguration.value = hostConfiguration
   }
-  LaunchedEffect(treehouseApp, treehouseView) {
-    treehouseApp.renderTo(treehouseView)
-  }
-  LaunchedEffect(treehouseView, contentSource) {
-    treehouseView.stateChangeListener?.onStateChanged(treehouseView)
+  DisposableEffect(treehouseView, contentSource) {
+    val closeable = contentSource.bindWhenReady(treehouseView, treehouseApp)
+    return@DisposableEffect object : DisposableEffectResult {
+      override fun dispose() = closeable.close()
+    }
```

Comment:
```suggestion
    onDispose {
      closeable.close()
    }
```

Reply:
Merged. I did it through Android Studio to remove the unnecessary import.